File: app/src/main/java/com/fastaccess/provider/timeline/handler/drawable/DrawableGetter.java
package com.fastaccess.provider.timeline.handler.drawable;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.text.Html;
import android.widget.TextView;

import com.bumptech.glide.GenericRequestBuilder;
import com.bumptech.glide.Glide;
import com.fastaccess.R;

import java.lang.ref.WeakReference;
import java.util.HashSet;
import java.util.Set;

/**
 * Created by Kosh on 22 Apr 2017, 7:44 PM
 */

public class DrawableGetter implements Html.ImageGetter, Drawable.Callback {
    private WeakReference<TextView> container;
    private final Set<GlideDrawableTarget> cachedTargets;
    private final int width;

    public DrawableGetter(TextView tv, int width) {
        tv.setTag(R.id.drawable_callback, this);
        this.container = new WeakReference<>(tv);
        this.cachedTargets = new HashSet<>();
        this.width = width;
    }

    @Override public Drawable getDrawable(@NonNull String url) {
        final UrlDrawable urlDrawable = new UrlDrawable();
        if (container != null && container.get() != null) {
            Context context = container.get().getContext();
            final GenericRequestBuilder load = Glide.with(context)
                    .load(url)
                    .placeholder(ContextCompat.getDrawable(context, R.drawable.ic_image))
                    .dontAnimate();
            final GlideDrawableTarget target = new GlideDrawableTarget(urlDrawable, container, width);
            load.override(width, width / 2);
            load.into(target);
            cachedTargets.add(target);
        }
        return urlDrawable;
    }

    @Override public void invalidateDrawable(@NonNull Drawable drawable) {
        if (container != null && container.get() != null) {
            container.get().invalidate();
        }
    }

    @Override public void scheduleDrawable(@NonNull Drawable drawable, @NonNull Runnable runnable, long l) {}

    @Override public void unscheduleDrawable(@NonNull Drawable drawable, @NonNull Runnable runnable) {}

    public void clear(@NonNull DrawableGetter drawableGetter) {
        if (drawableGetter.cachedTargets != null) {
            for (GlideDrawableTarget target : drawableGetter.cachedTargets) {
                Glide.clear(target);
            }
        }
    }
}

File: markdown/src/main/java/com/fastaccess/markdown/spans/drawable/DrawableGetter.kt
package com.fastaccess.markdown.spans.drawable

import android.graphics.drawable.Drawable
import android.text.Html
import android.widget.TextView
import androidx.annotation.NonNull
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.bumptech.glide.request.RequestOptions
import com.fastaccess.github.extensions.getDrawableCompat
import com.fastaccess.markdown.R
import timber.log.Timber

/**
 * Created by Kosh on 22 Apr 2017, 7:44 PM
 */

class DrawableGetter(
    private val tv: TextView?,
    private val width: Int,
    private val url: String
) : Html.ImageGetter, Drawable.Callback {
    private val cachedTargets = hashSetOf<GlideDrawableTarget<out Drawable>>()

    init {
        tv?.setTag(R.id.drawable_callback, this)
    }

    override fun getDrawable(oriUrl: String): Drawable {
        val urlDrawable = UrlDrawable()
        tv?.let {
            val context = it.context ?: return urlDrawable
            val imageTarget = if (oriUrl.endsWith(".gif")) {
                GlideDrawableTarget<GifDrawable>(urlDrawable, tv, width)
            } else {
                GlideDrawableTarget<Drawable>(urlDrawable, tv, width)
            }
            Glide.with(it).apply {
                clear(it)
                applyDefaultRequestOptions(RequestOptions().diskCacheStrategy(DiskCacheStrategy.ALL)
                    .placeholder(context.getDrawableCompat(R.drawable.ic_image)))
                if (oriUrl.endsWith(".gif")) {
                    asGif().load(url).into(imageTarget as GlideDrawableTarget<GifDrawable>)
                } else {
                    asDrawable().load(url).into(imageTarget as GlideDrawableTarget<Drawable>)
                }
            }
            cachedTargets.add(imageTarget)
        }
        return urlDrawable
    }

    override fun invalidateDrawable(@NonNull drawable: Drawable) {
        tv?.invalidate()
    }

    override fun scheduleDrawable(@NonNull drawable: Drawable, @NonNull runnable: Runnable, l: Long) {}

    override fun unscheduleDrawable(@NonNull drawable: Drawable, @NonNull runnable: Runnable) {}

    fun clear(@NonNull drawableGetter: DrawableGetter) {
        Timber.e("clearing......")
        for (target in drawableGetter.cachedTargets) {
            tv?.let { Glide.with(it).clear(target) }
        }
    }
}
